substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,770 @@
+	/// functions defined in the following impl block should act as a good reference for how the
+	/// exposed calls would look like when exposed.
+	impl<T: Config> Pallet<T> {
+		/// Register an account to become a stake `Agent`. Sometimes also called a `Delegatee`.
+		///
+		/// Delegators can authorize `Agent`s to stake on their behalf by delegating their funds to
+		/// them. The `Agent` can then use the delegated funds to stake to [`Config::CoreStaking`].
+		///
+		/// Implementation note: This function allows any account to become an agent. It is
+		/// important though that accounts that call [`StakingUnsafe::virtual_bond`] are keyless
+		/// accounts. This is not a problem for now since this is only used by other pallets in the
+		/// runtime which use keyless account as agents. If we later want to expose this as a
+		/// dispatchable call, we should derive a sub-account from the caller and use that as the
+		/// agent account.
+		pub fn register_agent(
+			origin: OriginFor<T>,
+			reward_account: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// Existing `agent` cannot register again.
+			ensure!(!Self::is_agent(&who), Error::<T>::NotAllowed);
+
+			// A delegator cannot become an `agent`.
+			ensure!(!Self::is_delegator(&who), Error::<T>::NotAllowed);
+
+			// They cannot be already a direct staker in the staking pallet.
+			ensure!(Self::not_direct_staker(&who), Error::<T>::AlreadyStaking);
+
+			// Reward account cannot be same as `agent` account.
+			ensure!(reward_account != who, Error::<T>::InvalidRewardDestination);
+
+			Self::do_register_agent(&who, &reward_account);
+			Ok(())
+		}
+
+		/// Migrate from a `Nominator` account to `Agent` account.
+		///
+		/// The origin needs to
+		/// - be a `Nominator` with `CoreStaking`,
+		/// - not already an `Agent`,
+		/// - have enough funds to transfer existential deposit to a delegator account created for
+		///   the migration.
+		///
+		/// This function will create a proxy account to the agent called `proxy_delegator` and
+		/// transfer the directly staked amount by the agent to it. The `proxy_delegator` delegates
+		/// the funds to the origin making origin an `Agent` account. The real `delegator`
+		/// accounts of the origin can later migrate their funds using [Self::claim_delegation] to
+		/// claim back their share of delegated funds from `proxy_delegator` to self.
+		pub fn migrate_to_agent(
+			origin: OriginFor<T>,
+			reward_account: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			// ensure who is not already an agent.
+			ensure!(!Self::is_agent(&who), Error::<T>::NotAllowed);
+
+			// and they should already be a nominator in `CoreStaking`.
+			ensure!(Self::is_direct_nominator(&who), Error::<T>::NotAllowed);
+
+			// Reward account cannot be same as `agent` account.
+			ensure!(reward_account != who, Error::<T>::InvalidRewardDestination);
```

Comment:
As commented above, maybe we should make sure that `reward_account != who && !Self::is_agent(reward_account)`.

Reply:
Some account can become agent first and reward account later.